File: src/sentiment/quantifier.py
"""
情绪量化引擎
"""

import os
import json
import math
from typing import Dict, List

# 来源可信度权重
SOURCE_WEIGHTS = {
    'yahoo_finance': 1.0,
    'eastmoney': 0.9,
    'sina': 0.8,
    'gdelt': 0.7,
    'xueqiu': 0.85,
}

# 情绪关键词
POSITIVE_WORDS = ['上涨', '增长', '盈利', '突破', '新高', '利好', 'surge', 'gain', 'profit', 'bullish']
NEGATIVE_WORDS = ['下跌', '亏损', '暴跌', '风险', '裁员', 'drop', 'loss', 'bearish', 'risk']


def analyze_sentiment(text: str) -> str:
    """分析文本情绪"""
    if not text:
        return 'neutral'
    
    text_lower = text.lower()
    
    pos_score = sum(1 for w in POSITIVE_WORDS if w in text_lower)
    neg_score = sum(1 for w in NEGATIVE_WORDS if w in text_lower)
    
    if pos_score > neg_score:
        return 'positive'
    elif neg_score > pos_score:
        return 'negative'
    else:
        return 'neutral'


def calc_news_score(news: Dict) -> float:
    """计算单条新闻的情绪分数"""
    sentiment = analyze_sentiment(news.get('title', '') + ' ' + news.get('content', ''))
    
    if sentiment == 'positive':
        base = 50
    elif sentiment == 'negative':
        base = -50
    else:
        base = 0
    
    # 来源可信度
    source = news.get('source', 'unknown')
    for key, weight in SOURCE_WEIGHTS.items():
        if key in source.lower():
            return base * weight
    
    return base * 0.7


def quantify_sentiment(symbol: str = None) -> Dict:
    """
    量化情绪分数 (-100 到 +100)
    扫描 news_items 表中的新闻
    """
    # 尝试读取数据库
    try:
        import subprocess
        
        db_url = os.getenv('DATABASE_URL', '')
        
        # 使用 psql 或直接读取缓存
        cache_file = 'data/sentiment_cache.json'
        if os.path.exists(cache_file):
            with open(cache_file) as f:
                data = json.load(f)
            
            # 计算情绪
            total = 0
            for item in data.get('items', []):
                total += calc_news_score(item)
            
            score = max(-100, min(100, total))
            
            return {
                'score': round(score, 1),
                'count': len(data.get('items', [])),
                'positive': sum(1 for i in data.get('items', []) if i.get('sentiment') == 'positive'),
                'negative': sum(1 for i in data.get('items', []) if i.get('sentiment') == 'negative'),
                'neutral': sum(1 for i in data.get('items', []) if i.get('sentiment') == 'neutral'),
                'sources': data.get('sources', {}),
                'latest_news': data.get('items', [])[:5]
            }
    except Exception as e:
        print(f"Error: {e}")
    
    # 如果数据库不可用，返回模拟数据
    return {
        'score': 0,
        'count': 0,
        'positive': 0,
        'negative': 0,
        'neutral': 0,
        'sources': {},
        'latest_news': []
    }


def calculate_momentum(symbol: str = None) -> Dict:
    """计算情绪变化率"""
    sent = quantify_sentiment(symbol)
    
    if sent['score'] > 10:
        trend = 'improving'
    elif sent['score'] < -10:
        trend = 'declining'
    else:
        trend = 'stable'
    
    return {
        'current': sent['score'],
        'previous': 0,
        'momentum': 0,
        'trend': trend
    }


# 测试
if __name__ == '__main__':
    import sys
    symbol = sys.argv[1] if len(sys.argv) > 1 else None
    
    result = quantify_sentiment(symbol)
    print(f"情绪分数: {result['score']}")
    print(f"新闻数量: {result['count']}")
    print(f"正面: {result['positive']}, 负面: {result['negative']}")

File: src/sentiment/quantifier.ts
/**
 * 情绪量化模块
 * 将新闻情绪量化为可交易的分数 (-100 到 +100)
 */

import pg from 'pg';

const { Pool } = pg;

let pool: any;

function getPool() {
  if (!pool) {
    pool = new Pool({
      connectionString: process.env.DATABASE_URL || 'postgresql://zhengzefeng:password@localhost:5432/trading_bot'
    });
  }
  return pool;
}

// 来源可信度
const SOURCE_CREDIBILITY: Record<string, number> = {
  'yahoo_finance': 1.0,
  'yahoo_finance_us': 1.0,
  'eastmoney': 0.9,
  'sina_finance': 0.8,
  'gdelt': 0.7,
  'xueqiu': 0.85,
  'default': 0.7,
};

// 默认可信度
const DEFAULT_CREDIBILITY = 0.7;

// 时效性衰减参数
const DECAY_HALF_LIFE_HOURS = 24;
const DECAY_FACTOR = 0.5;

/**
 * 获取新闻
 */
async function getNewsForSymbol(symbol: string, hours: number = 72): Promise<any[]> {
  const result = await getPool().query(`
    SELECT id, title, source, published_at, sentiment, market
    FROM news_items
    WHERE (title LIKE $1 OR content LIKE $1 OR symbols LIKE $1)
    AND published_at > NOW() - INTERVAL '${hours} hours'
    ORDER BY published_at DESC
    LIMIT 100
  `, [`%${symbol}%`]);
  
  return result.rows;
}

/**
 * 获取全局新闻 (不限定symbol)
 */
async function getGlobalNews(hours: number = 72): Promise<any[]> {
  const result = await getPool().query(`
    SELECT id, title, source, published_at, sentiment, market
    FROM news_items
    WHERE published_at > NOW() - INTERVAL '${hours} hours'
    ORDER BY published_at DESC
    LIMIT 200
  `);
  
  return result.rows;
}

/**
 * 计算单条新闻的情绪贡献
 */
function calculateNewsSentiment(news: any): number {
  // 基础分数
  let score = 0;
  
  switch (news.sentiment) {
    case 'positive':
      score = 50;
      break;
    case 'negative':
      score = -50;
      break;
    case 'neutral':
    default:
      score = 0;
      break;
  }
  
  // 来源可信度
  const credibility = SOURCE_CREDIBILITY[news.source] || DEFAULT_CREDIBILITY;
  score *= credibility;
  
  return score;
}

/**
 * 计算时效性衰减
 */
function getTimeDecay(publishedAt: Date): number {
  const now = new Date();
  const hoursAgo = (now.getTime() - new Date(publishedAt).getTime()) / (1000 * 60 * 60);
  
  if (hoursAgo <= 0) {
    return 1.0;
  }
  
  // 指数衰减
  return Math.pow(DECAY_FACTOR, hoursAgo / DECAY_HALF_LIFE_HOURS);
}

/**
 * 量化情绪分数 (-100 到 +100)
 */
export async function quantifySentiment(symbol?: string): Promise<{
  score: number;
  count: number;
  positive: number;
  negative: number;
  neutral: number;
  sources: Record<string, number>;
  latest_news: any[];
}> {
  const news = symbol 
    ? await getNewsForSymbol(symbol)
    : await getGlobalNews();
  
  if (news.length === 0) {
    return {
      score: 0,
      count: 0,
      positive: 0,
      negative: 0,
      neutral: 0,
      sources: {},
      latest_news: [],
    };
  }
  
  let totalScore = 0;
  let positive = 0;
  let negative = 0;
  let neutral = 0;
  const sourceCounts: Record<string, number> = {};
  const latestNews: any[] = [];
  
  for (const n of news) {
    // 时效性衰减
    const decay = getTimeDecay(n.published_at);
    
    // 计算情绪贡献
    const newsScore = calculateNewsSentiment(n) * decay;
    totalScore += newsScore;
    
    // 统计
    if (n.sentiment === 'positive') {
      positive++;
    } else if (n.sentiment === 'negative') {
      negative++;
    } else {
      neutral++;
    }
    
    // 来源统计
    sourceCounts[n.source] = (sourceCounts[n.source] || 0) + 1;
    
    // 保存最新新闻
    if (latestNews.length < 5) {
      latestNews.push({
        title: n.title,
        source: n.source,
        sentiment: n.sentiment,
        published_at: n.published_at,
        score: newsScore,
      });
    }
  }
  
  // 归一化到 -100 到 +100
  const normalizedScore = Math.max(-100, Math.min(100, totalScore));
  
  return {
    score: normalizedScore,
    count: news.length,
    positive,
    negative,
    neutral,
    sources: sourceCounts,
    latest_news: latestNews,
  };
}

/**
 * 计算情绪变化率 (momentum)
 */
export async function calculateMomentum(symbol?: string, periods: number = 5): Promise<{
  current: number;
  previous: number;
  momentum: number;
  trend: 'improving' | 'declining' | 'stable';
}> {
  // 获取多个时间段的情绪
  const scores: number[] = [];
  
  for (const hours of [24, 48, 72, 96, 120]) {
    const sentiment = symbol 
      ? await quantifySentiment(symbol)
      : await quantifySentiment();
    scores.push(sentiment.score);
  }
  
  if (scores.length < 2) {
    return {
      current: scores[0] || 0,
      previous: 0,
      momentum: 0,
      trend: 'stable',
    };
  }
  
  const current = scores[0];
  const previous = scores[Math.min(periods - 1, scores.length - 1)];
  const momentum = current - previous;
  
  let trend: 'improving' | 'declining' | 'stable' = 'stable';
  if (momentum > 10) {
    trend = 'improving';
  } else if (momentum < -10) {
    trend = 'declining';
  }
  
  return {
    current,
    previous,
    momentum,
    trend,
  };
}

/**
 * 获取市场的整体情绪
 */
export async function getMarketSentiment(): Promise<{
  score: number;
  momentum: number;
  trend: 'improving' | 'declining' | 'stable';
  symbols: Record<string, number>;
}> {
  const sentiment = await quantifySentiment();
  const momentumData = await calculateMomentum();
  
  // 统计各 symbol 的情绪
  const symbolScores: Record<string, number> = {};
  
  // 从 news_items 获取有 symbol 标签的新闻
  const result = await pool.query(`
    SELECT symbols, sentiment 
    FROM news_items
    WHERE published_at > NOW() - INTERVAL '24 hours'
    AND symbols IS NOT NULL
    LIMIT 100
  `);
  
  for (const row of result.rows) {
    if (!row.symbols) continue;
    
    const symbols = row.symbols.split(',').map((s: string) => s.trim());
    const score = row.sentiment === 'positive' ? 1 : row.sentiment === 'negative' ? -1 : 0;
    
    for (const sym of symbols) {
      symbolScores[sym] = (symbolScores[sym] || 0) + score;
    }
  }
  
  return {
    score: sentiment.score,
    momentum: momentumData.momentum,
    trend: momentumData.trend,
    symbols: symbolScores,
  };
}

// 测试
if (require.main === module) {
  quantifySentiment().then(result => {
    console.log('情绪分数:', result.score);
    console.log('新闻数量:', result.count);
    console.log('正面:', result.positive, '负面:', result.negative, '中性:', result.neutral);
    pool.end();
  }).catch(console.error);
}

export default quantifySentiment;